Use FreeIPA API to authenticate and authorize users instead of LDAP.

felddy opened the pull request:
# <!--- Provide a general summary of your changes in the Title above -->

## 🗣 Description

<!--- Describe your changes in detail -->
- Replace LDAP calls with kerberized calls to the FreeIPA API.
- Compare user TLS certificates instead of just their CN.
- Modernize role.
- Upstream changes.

## 💭 Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

We were running into issues with LDAP comparing DNs.  Specifically multi-value RDNs.  The order of AVAs with and RDNs is insignificant (is a `set`).  There is no way to "normalize" a CN that will thwart our badging offices ability to scramble up our UIDs nicely.  

See: https://frasertweedale.github.io/blog-redhat/posts/2019-05-28-a-dn-is-not-a-string.html

We are now configuring OpenVPN to pass the user's entire public certificate to us as a file.  It is then sent to FreeIPA for matching.

As an added benefit we can now ask FreeIPA interesting questions not only about group membership, but multiple attributes that would deactivate an account, preventing login. 

## 🧪 Testing

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Tested in staging.
Molecule tests.

## 📷 Screenshots (if appropriate)

## 🚥 Types of Changes

<!--- What types of changes does your code introduce? -->
<!--- Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (causes existing functionality to change)

## ✅ Checklist

<!--- Go over all the following points, and put an `x` in all the
boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask.
We're here to help! -->

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] Big thanks to @jsf9k for fixing things I could not.

felddy committed:
Convert user verification script to use FreeIPA API instead of LDAP

felddy committed:
Update ansible task to install new libraries

felddy committed:
Add kerberos development library to Debian

felddy committed:
Add Kerberos development libraries to Fedora

jsf9k committed:
Move verify-cn tasks to a separate file

This helps keep the main tasks file to a manageable size.

jsf9k committed:
Add verify-cn variables for Ubuntu

Also note that Ubuntu 16.04 calls the OpenVPN service by a different
name.

jsf9k committed:
Add verify-cn variables for RedHat and Debian

Also remove the vars file specific to Debian 9.  It is no longer
required.

felddy committed:
Disable PrivateTmp in OpenVPN system description

PrivateTmp is preventing OpenVPN from sharing user certificates with the 
verify-cn.py child process.

felddy committed:
Merge branch 'improvement/use_freeipa_api_2' of github.com:cisagov/ansible-role-openvpn into improvement/use_freeipa_api_2

felddy committed:
Merge branch 'develop' of github.com:cisagov/ansible-role-openvpn into improvement/use_freeipa_api_2

felddy committed:
Fix typo in comment

jsf9k commented on vars/Debian.yml:
```diff
@@ -20,3 +19,14 @@ iptables_persistence_package_names:
+# script
+verify_cn_pip_package_names:
+  - python-freeipa
+  - requests-gssapi
```
I still question whether requests-gssapi is required for any distribution.  I don't see it imported or used anywhere.

felddy replied:
This is required to use the `login_kerberos` function of `python-freeipa`.
See: https://python-freeipa.readthedocs.io/en/latest/#python_freeipa.client.Client.login_kerberos

jsf9k replied:
Got it!

jsf9k reviewed (changes requested):
I will approve once you have addressed the concern I have raised before.  Please see my comment.

dav3r commented on files/verify-cn.py:
```diff
@@ -19,94 +20,152 @@
+    # Create the FreeIPA client
+    ipa_client: ClientMeta = ClientMeta(dns_discovery=realm)
+
+    # Login client in using kerberos credentials
```
There's an extra `"in"` in this comment- pick one to keep and one to toss.

dav3r replied:
Fixed in https://github.com/cisagov/ansible-role-openvpn/pull/18/commits/32f83bde13b8778d07ca55771f7a668de2093ef1 👍

dav3r commented on files/verify-cn.py:
```diff
@@ -19,94 +20,152 @@
-        logging.warning(f"More than 1 record found in LDAP.  Count: {len(rec)}.")
+    if response["count"] > 1:
+        logging.warning(
+            f"More than 1 user matched certificate.  Count: {response['count']}."
```
Can we log all of the UIDs found in this case?  That would make resolving this situation much easier.

felddy replied:
Good idea. 
See: c065295

dav3r replied:
The future admin who encounters this situation thanks you sir!  🍹

dav3r commented on files/verify-cn.py:
```diff
@@ -19,94 +20,152 @@
+        logging.info(f"{matched_uid} will be permitted access.")
+        return True
+    else:
+        logging.warning(f"{matched_uid} ACCESS DENIED.")
```
⛔⛔⛔ 👍

jsf9k replied:
Previously @mcdonnnj pointed out that there are good reasons to prefer format strings to f-strings in `logging` calls like this.  Not sure if you had forgotten this or were purposefully disregarding it.  We should as a team choose one style and use it everywhere.

dav3r replied:
Forgot about that- we should avoid f-strings in logging statements.

https://blog.pilosus.org/posts/2020/01/24/python-f-strings-in-logging/

dav3r replied:
Begrudgingly adjusted by @felddy in https://github.com/cisagov/ansible-role-openvpn/pull/18/commits/3b0a4f629408e4c26f5227eee74117e2705f3415. 😜

felddy replied:
😏

dav3r commented on files/verify-cn.yml:
```diff
@@ -1,3 +1,4 @@
 ---
-realm: COOL.CYBER.DHS.GOV
+log_level: INFO
+realm: STAGING.COOL.CYBER.DHS.GOV
```
Is this `STAGING.` supposed to be here?

felddy replied:
yes. This file is just a placeholder.  The real file is created by cloud-init.

dav3r reviewed (commented):
In addition to @jsf9k's important question, I added a few minor comments/questions.

dav3r commented on files/verify-cn.py:
```diff
@@ -19,94 +20,152 @@
+        client_certificate_path: Path = Path(os.environ[PEER_CERT_VARIABLE])
+    except KeyError:
+        logging.critical(
+            "OpenVPN did not set the 'peer_cert' environement variable."
```
Typo: `environement`

dav3r replied:
Fixed in https://github.com/cisagov/ansible-role-openvpn/pull/18/commits/79c9a9199c6ee6436362e3181db86c6c46377c22 👍

felddy committed:
Add missing pip package requirement

For DNS service discovery, you need to have the srvlookup module 
installed.

https://python-freeipa.readthedocs.io/en/latest/

felddy committed:
Fix typo in logging message

felddy committed:
Convert logging f-strings to printf-style

See: 
https://blog.pilosus.org/posts/2020/01/24/python-f-strings-in-logging/

felddy committed:
Make English better

felddy committed:
Modify OpenVPN configuration to store client certificates temporarily

jsf9k commented on files/primary.conf:
```diff
@@ -313,6 +313,13 @@ verb 3
 # can automatically reconnect.
 explicit-exit-notify 1
 
+# Store the certificates the clients uses upon connection to this directory.
```
"clients uses" is _poor_ English.

felddy replied:
My bad.  Copied from man page without proofreading it.

felddy committed:
Fix description copied from man page

felddy committed:
Add log messages when multiple users match a certificate

felddy committed:
Fix bad assumption about "count" response, and log multi-user matches

jsf9k reviewed (approved):
:zap:

dav3r reviewed (approved):
This is some 🌶️ 💩 right here!  🚀
